### PR TITLE
Fixed static typing bug where app.template_global() doesn't accept arguments #4098

### LINKED ISSUE #4098: Typing requires template globals to accept no arguments
The typings in Flask 2.0.1 require a function decorated with `@app.template_global()` to take no arguments. As far as I’m aware, template globals certainly are supposed to take arguments (e.g. the [example in the API documentation](https://flask.palletsprojects.com/en/2.0.x/api/#flask.Flask.template_global) takes one argument), so I assume this is not intended. I think this specific issue was introduced with #4071, which applied the `TemplateGlobalCallable` type alias to that decorator; however, the `TemplateGlobalCallable` type itself is already an alias for `t.Callable[[], t.Any]` (i.e., take no arguments and return anything) in Flask 2.0.0. I haven’t checked yet if `app.add_template_global()` is affected.

Example code:

```python
import flask


app = flask.Flask(__name__)


@app.template_global()
def identity(arg: str) -> str:
    return arg
```

> app.py:7: error: Argument 1 has incompatible type "Callable[[str], str]"; expected "Callable[[], Any]"
> Found 1 error in 1 file (checked 1 source file)

Sample repository: https://github.com/lucaswerkmeister/repro-flask-mypy-template_global

Steps to reproduce:

```sh
git clone https://github.com/lucaswerkmeister/repro-flask-mypy-template_global
make -C repro-flask-mypy-template_global
```

Environment:

- Python version: 3.9.5
- Flask version: 2.0.1
- Mypy version: 0.812

### PR DESCRIPTION
**Pull request by default-303**
Checklist:

- [x] Add an entry in `CHANGES.rst` summarizing the change and linking to the issue.
- [x] Run `pre-commit` hooks and fix any issues.
- [x] Run `pytest` and `tox`, no tests failed.

Fixes this issue mentioned in #4098 where static type checker like `mypy` would raise an error if a `function` decorated by `global template` methods/decorators like `app.template_global()` requires an argument. 

fixes #4098

### CODE REVIEW

**Comment by davidism** [comment]
Might as well fix the filter and test callables too.

**Comment by default-303** [comment]
hey, I'm kinda new so can you elaborate what you expect me to do ?

**Comment by davidism** [comment]
There are two other callable types around the one you modified that have the same problem you're fixing.

**Comment by default-303** [comment]
hey, can you tell if you are happy with the change or it needs a further review  ?

**Comment by Yourun-proger** [comment]
@default-303 , most likely, the matter is that you have not added tests confirming the correct execution of the code.

**Comment by ThiefMaster** [comment]
@Yourun-proger I somehow doubt that this PR needs unit tests. It's a typing change...

**Comment by davidism** [comment]
@Yourun-proger please, I appreciate you trying to help, but leave it to maintainers and experienced contributors to help first time contributors. In this case, your advice doesn't make sense, this is a change to typing, there's no tests to add.

File: CHANGES.rst [hunk 1]
```diff
@@ -7,6 +7,12 @@ Unreleased
 
 -   Update Click dependency to >= 8.0.
 
+-   Fixed the issue where typing requires ``template global``
+    decorators to accept functions with no arguments. This meant
+    user defined functions which require to take in arguments
+    fail the lint tests (eg - ``mypy``).
+    Fixes ``#4098``
```
Comment (davidism) [inline CHANGES.rst:14]: Please follow the style used in the rest of the changelog for linking to issues. Also, can remove the section describing that the incorrect type caused type tests to fail, only the first sentence is needed. Also, `template global` is a phrase, not code, so it shouldn't be in code formatting.

**Comment by default-303** [comment]
Okay, I have no problem adding tests but I   don't think you can add tests for static typing.
That said I did search but couldn't find any resources.

Consider adding them if you know of any

**Comment by davidism** [comment]
You don't need to add tests, please ignore that user.

**Comment by davidism** [comment]
Since this is a bug fix, please rebase this to the 2.0.x maintenance branch. `git rebase -i origin/2.0.x` then `git push -f`.

**Comment by default-303** [comment]
hey, I've squashed all the commits. 
Check and let me know if I need to do anything else.

**Comment by default-303** [comment]
@davidism can i get some sort of update ?

**Comment by ThiefMaster** [comment]
I don't think your rebase went well. If you look at the commits, there's 2.1 stuff (from main) in it...

File: src/flask/typing.py [hunk 2]
```diff
@@ -39,8 +39,8 @@
 ErrorHandlerCallable = t.Callable[[Exception], ResponseReturnValue]
 TeardownCallable = t.Callable[[t.Optional[BaseException]], "Response"]
 TemplateContextProcessorCallable = t.Callable[[], t.Dict[str, t.Any]]
-TemplateFilterCallable = t.Callable[[t.Any], str]
-TemplateGlobalCallable = t.Callable[[], t.Any]
-TemplateTestCallable = t.Callable[[t.Any], bool]
+TemplateFilterCallable = t.Callable[..., str]
```
Comment (ThiefMaster) [inline src/flask/typing.py:42]: I think the return value here should be `t.Any` as well..

**Comment by default-303** [comment]
So you want me to do a pr from 2.0.x branch instead of 'typing_hotfix'?

**Comment by default-303** [comment]
> I don't think your rebase went well. If you look at the commits, there's 2.1 stuff (from main) in it...

So you want me to do a pr from 2.0.x branch instead of 'typing_hotfix'?

**Comment by ThiefMaster** [comment]
No, simply rebase your branch so it only has your commit on top of 2.0.x and then force-push. You do not need to open a new PR for this.

**Comment by ThiefMaster** [comment]
`git rebase --onto origin/2.0.x main` should do the job.

**Comment by default-303** [comment]
God, did everything go the way you guys do things here ? Hopefully yes

File: CHANGES.rst [hunk 3]
```diff
@@ -1,5 +1,15 @@
 .. currentmodule:: flask
 
+Version 2.1.0
+-------------
+
+Unreleased
+
+-   Update Click dependency to >= 8.0.
```
Comment (ThiefMaster) [inline CHANGES.rst:8]: Remove this and the rest of the 2.1.0 stuff

File: CHANGES.rst [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+-   Fixed the issue where typing requires template global
+    decorators to accept functions with no arguments. :issue:`4098`
```
Comment (ThiefMaster) [inline CHANGES.rst:10]: This should be part of the 2.0.2 changelog

**Comment by ThiefMaster** [comment]
almost there :) just some fixes in the changelog

File: CHANGES.rst (hunk 4 above)
Comment (default-303) [reply CHANGES.rst]: sure

**Comment by default-303** [comment]
that should do it (yay!!)

**Pull request merged**